src/batch.rs

```diff
@@ -0,0 +1,297 @@
+    /// - R_i is the signature's R value;
+    /// - s_i is the signature's s value;
+    /// - c_i is the hash of the message and other data;
+    /// - z_i is a random 128-bit Scalar.
+    /// - h_G is the cofactor of the group;
+    ///
+    /// As follows elliptic curve scalar multiplication convention,
+    /// scalar variables are lowercase and group point variables
+    /// are uppercase. This does not exactly match the RedDSA
+    /// notation in the [protocol specification §B.1][ps].
+    ///
+    /// [ps]: https://zips.z.cash/protocol/protocol.pdf#reddsabatchverify
+    #[allow(non_snake_case)]
+    pub fn verify<R: RngCore + CryptoRng>(self, mut rng: R) -> Result<(), Error> {
+        let n = self.signatures.len();
+
+        let mut VK_coeffs = Vec::with_capacity(n);
+        let mut VKs = Vec::with_capacity(n);
+        let mut R_coeffs = Vec::with_capacity(self.batch_size);
+        let mut Rs = Vec::with_capacity(self.batch_size);
+        let mut P_spendauth_coeff = Scalar::zero();
+        let mut P_binding_coeff = Scalar::zero();
+
+        for item in self.signatures.iter() {
+            match item.inner {
+                Inner::SpendAuth { vk_bytes, sig, c } => {
+                    // let tup: BatchTuple<SpendAuth> = (vk_bytes, sig, c);
+
+                    // let (P_coeff, R, R_coeff, VK, VK_coeff) =
+                    //     &self.compute_multiscalar_mul_inputs(rng, tup)?;
+
+                    let z = Scalar::from_raw(gen_128_bits(&mut rng));
+
+                    let s = {
+                        // XXX-jubjub: should not use CtOption here
+                        let maybe_scalar = Scalar::from_bytes(&sig.s_bytes);
+                        if maybe_scalar.is_some().into() {
+                            maybe_scalar.unwrap()
+                        } else {
+                            return Err(Error::InvalidSignature);
+                        }
+                    };
+
+                    let R = {
+                        // XXX-jubjub: should not use CtOption here
+                        // XXX-jubjub: inconsistent ownership in from_bytes
+                        let maybe_point = AffinePoint::from_bytes(sig.r_bytes);
+                        if maybe_point.is_some().into() {
+                            jubjub::ExtendedPoint::from(maybe_point.unwrap())
+                        } else {
+                            return Err(Error::InvalidSignature);
+                        }
+                    };
+                    let P_coeff = z * s;
+                    let R_coeff = z;
+                    let VK = VerificationKey::<SpendAuth>::try_from(vk_bytes.bytes)
+                        .unwrap()
+                        .point;
+                    let VK_coeff = Scalar::zero() + (z * c);
+
+                    P_spendauth_coeff -= P_coeff;
+                    Rs.push(R);
+                    R_coeffs.push(R_coeff);
+                    VKs.push(VK);
+                    VK_coeffs.push(VK_coeff);
```

Comment:
@hdevalence @yaahc I tried moving this highly !DRY body into a method on Verifier but I way having issues with `Borrow` once everything was computed and I was trying to hand iterators to `vartime_mutiscalar_mul` at the bottom. Advice on making this DRYer and nicer is much appreciated.

Reply:
The verification code is identical in both cases, except that: (1) some of the types are different (but only because of marker parameters) and (2) it accumulates into a different coefficient `P_spendauth_coeff` or `P_binding_coeff`.

I think it's possible to have one set of code, using inline match statements to handle the tiny divergences.  For instance,

```
let (s_bytes, r_bytes, c) = match item.inner {
    Inner::SpendAuth { sig, c, .. } => (&sig.s_bytes, &sig.r_bytes, &c),
    Inner::Binding { sig, c, .. } => (&sig.s_bytes, &sig.r_bytes, &c),
};
```
and
```
let vk = match item.inner {
    Inner::SpendAuth { vk_bytes, .. } => VerificationKey::<SpendAuth>::try_from(vk_bytes.bytes)?.point,
    Inner::Binding { vk_bytes, .. } => VerificationKey::<Binding>::try_from(vk_bytes.bytes)?.point,
};
```
and
```
match item.inner {
    Inner::SpendAuth { .. } => {
        P_spendauth_coeff -= P_coeff,
    },
    Inner::Binding { .. } => {
        P_binding_coeff -= P_coeff,
    },
}
```
Then the rest of the code should be identical.